integreat_compass/cms/views/index.py

```diff
@@ -1,9 +1,124 @@
+import re
+
+from django.shortcuts import render
 from django.views.generic import TemplateView
 
+from ..constants import offer_group_types, offer_mode_types
+from ..models import Offer, Tag
+
 
 class IndexView(TemplateView):
     """
-    Start page of the project
+    View to see all offers
     """
 
     template_name = "index.html"
+
+    def get(self, request):
+        """
+        This functions displays all offers
+
+        :param request: The current request
+        :type request: ~django.http.HttpRequest
+
+        :return: The rendered template response
+        :rtype: ~django.template.response.TemplateResponse
+        """
+        offers = Offer.objects.all()
+        tags = Tag.objects.all()
+        list_of_offers = []
+        for offer in offers:
+            list_of_offers.append(self.store_offer_data_in_dict(offer))
+        amount_of_offers = len(list_of_offers)
+
+        return render(
+            request,
+            self.template_name,
+            {
+                "tags" : tags,
+                "offers" : list_of_offers,
+                "amount_of_offers" : amount_of_offers,
+                "offer_group_types" : offer_group_types.CHOICES,
+                "offer_mode_types" : offer_mode_types.CHOICES
+            },
+        )
+    
+    def post(self, request):
+        """
+        This functions filters
+
+        :param request: The current request
+        :type request: ~django.http.HttpRequest
+
+        :return: The rendered template response
+        :rtype: ~django.template.response.TemplateResponse
+        """
+        offers = Offer.objects.all()
+        tags = Tag.objects.all()
+
+        selected_filters = {
+            "tags" : request.POST.getlist("offer-tags"),
+            "mode_type" : request.POST.get("mode-type"),
+            "group_type" : request.POST.get("group-type"),
+            "search" : request.POST.get("offer-search"),
+        }
+        
+        filtered_offers = []
+        selected_filters["tags"] = [eval(i) for i in selected_filters["tags"]]
+
+
+        for offer in offers:
+            tag_ids = [tag.id for tag in offer.tags.all()]
+
+            if any(item in selected_filters["tags"] for item in tag_ids):
```

Comment:
This part is supposed to do the logic for filtering. However it's not fully functional yet - and also doesn't seem to be the best solution at all. All parts work individually, but they don't work together. Do you have an idea, how to create the filter logic if two or more filter criteria have been selected (for example online offer and has "Beruf" in the title)

Reply:
When filtering database objects in Django, it's usually best to do the filtering via the [queryset filters](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#filter).
This has the advantage that Django can [defer the execution](https://docs.djangoproject.com/en/4.2/topics/db/queries/#querysets-are-lazy) of the sql statements and only execute one query in the end which contains the final objects, so we don't retrieve objects from the database which we don't need anyway.

The only exception to this rule is when you e.g. need the intermediary results or perform other operations which intercept the lazy loading and cause multiple queries.

In this case, it depends on how the filtering should work - if multiple tags are selected, do we want all offers which contain at least one of these tags, or all of them?

The first one is a bit easier:
```
offers = offers.filter(tags__in=selected_filters["tags"]).distinct()
```

While the latter requires a loop (but due to the lazy execution of querysets, this only generates one query in the end):
```
for tag in selected_filters["tags"]:
    offers = offers.filter(tags=tag)
```

And this loop can also be written as fancy python syntax (but this only changes way the query is generated, it does not change the fact that the filtering happens on the sql layer):
```
from functools import reduce

offers = reduce(lambda offers, tag: offers.filter(tags=selected_filters["tags"]), offers)
```

In any case, you don't need to know the whole queryset documentation by heart, but you should know where to look whenever you face similar tasks in the future.

Reply:
Thank you very much for this detailed description. Turns out I confused myself just before asking the question. The bigger issue for me was how to stack filters on top of each other, but turns out we did it for Integreat too and it's quite simple. Sorry!